content/blog/2022-04-21-debug-your-lamda-functions-from-your-ide-using-localstack/index.md

```diff
@@ -0,0 +1,242 @@
+            "pathMappings": [
+                {
+                    "localRoot": "${workspaceFolder}",
+                    "remoteRoot": "."
+                }
+            ]
+        }
+    ]
+}
+```
+
+You have about 15 seconds to switch to Visual Studio Code and run the preconfigured remote debugger with the above function. This timeout is configurable, and now you can set a breakpoint in the Lambda handler code first, which can then later be inspected.
+
+Here is our Lambda debugger in action:
+
+{{< img src="lambda-function-debugging-vs-code.png" >}}
+
+## Configuring LocalStack for JVM debugging
+
+To configure LocalStack for JVM debugging, you would need to set the `LAMBDA_JAVA_OPTS` with jdwp settings and expose the debug port of your choice. Over a `docker-compose.yml`, you can do that via:
+
+```yml
+services:
+  localstack:
+    ...
+    environment:
+      ...
+      - LAMBDA_JAVA_OPTS=-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=*:5050
+      - LAMBDA_DOCKER_FLAGS=-p 127.0.0.1:5050:5050
+```
+
+The `suspend=y` option delays the code execution until debugger is attached to debgger server. To change the behavior, turn it to `suspend=n`. 
+
+## Configuring Visual Studio Code for JVM Lambda debugging
+
+To configure Visual Studio Code, for JVM Lambda debugging, we need to install the [Language support for Java](https://marketplace.visualstudio.com/items?itemName=redhat.java) and [Debugger for Java](https://marketplace.visualstudio.com/items?itemName=vscjava.vscode-java-debug) extensions. Create a new `.vscode/tasks.json` file, if not already present, and add a new task to it:
+
+```json
+{
+    "version": "2.0.0",
+    "tasks": [
+        {
+          "label": "Wait Remote Debugger Server",
+          "type": "shell",
+          "command": "while [[ -z $(docker ps | grep :5050) ]]; do sleep 1; done; sleep 1;"
+        }
+    ]
+}
+```
+
+Create a new `launch.json` file or edit an existing one from the Run and Debug tab, then add the following configuration:
+
+```json
+{
+    "version": "0.2.0",
+    "configurations": [
+        {
+            "type": "java",
+            "name": "Remote JVM on LS Debug",
+            "projectRoot": "${workspaceFolder}",
+            "request": "attach",
+            "hostName": "localhost",
+            "preLaunchTask": "Wait Remote Debugger Server",
+            "port": 5050
+        }
+    ]
+}
+```
+
+To debug your lambda function, click on the `Debug` icon with `Remote JVM on LS Debug` configuration selected, and then invoke your Lambda function.
+
+## Conclusion
+
+Debugging your Lambda function with LocalStack on your IDE is a great way to test and validate your Lambda functions before pushing them to production. Since you have already mounted the local Lambda function from your local filesystem into the Lambda container, LocalStack would immediately reflect the changes in the above section. Here is an example where you change the implementation of the handler as follows:
```

Comment:
```suggestion
Debugging your Lambda function with LocalStack in your IDE is a great way to enable quick feedback cycles, and to test and validate your code logic before pushing it to production. Since you have already mounted the local Lambda function from your local filesystem into the Lambda container, LocalStack would immediately reflect the changes in the above section. Here is an example where you change the implementation of the handler as follows:
```